sycl/test-e2e/Matrix/Inputs/joint_matrix_out_bounds_impl.hpp

```diff
@@ -83,42 +91,69 @@ void matrix_multiply(T1 *C, T2 *A, T2 *B, queue q) {
-  bfloat16 *vnniB = malloc_shared<bfloat16>(MATRIX_K * MATRIX_N, q);
-  float *C = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
-  float *D = malloc_shared<float>(MATRIX_M * MATRIX_N, q);
-
-  matrix_rand(MATRIX_M, MATRIX_K, A, (bfloat16)5);
-  matrix_rand(MATRIX_K, MATRIX_N, B, (bfloat16)5);
-  matrix_fill(MATRIX_M, MATRIX_N, C, (float)1);
-  matrix_fill(MATRIX_M, MATRIX_N, D, (float)1);
-
-  matrix_vnni<bfloat16>(MATRIX_K, MATRIX_N, B, vnniB, vnniFactor);
 
+  // reference data
+  Tab *A = malloc_shared<Tab>(MATRIX_M * MATRIX_K, q);
+  Tab *B = malloc_shared<Tab>(MATRIX_K * MATRIX_N, q);
+  Tc *C = malloc_shared<Tc>(MATRIX_M * MATRIX_N, q);
+  Tc *D = malloc_shared<Tc>(MATRIX_M * MATRIX_N, q);
+  matrix_rand(MATRIX_M, MATRIX_K, A, (Tab)5);
+  matrix_rand(MATRIX_K, MATRIX_N, B, (Tab)5);
+  matrix_fill(MATRIX_M, MATRIX_N, D, (Tc)1);
   matrix_multiply_ref(A, B, D, MATRIX_M, MATRIX_N, MATRIX_K);
-  matrix_multiply<float, bfloat16, MATRIX_M, MATRIX_K, MATRIX_K / vnniFactor,
-                  MATRIX_N * vnniFactor, MATRIX_M, MATRIX_N,
-                  layout::ext_intel_packed, vnniFactor>(C, A, vnniB, q);
-  bool res = matrix_compare(MATRIX_M, MATRIX_N, C, D);
-
-  matrix_multiply<float, bfloat16, MATRIX_M, MATRIX_K, MATRIX_K, MATRIX_N,
-                  MATRIX_M, MATRIX_N, layout::row_major, 1>(C, A, B, q);
-  res = res && matrix_compare(MATRIX_M, MATRIX_N, C, D);
 
-  std::cout << (res ? "passed" : "failed") << std::endl;
+  // test data
+  if constexpr (A_layout == layout::col_major) {
+    Tab *colA = malloc_shared<Tab>(MATRIX_K * MATRIX_M, q);
+    matrix_transpose(MATRIX_M, MATRIX_K, colA, A);
+    Tab *tmp = A;
+    A = colA;
+    free(tmp, q);
+  }
+
+  if constexpr (B_layout == layout::col_major) {
+    Tab *colB = malloc_shared<Tab>(MATRIX_N * MATRIX_K, q);
+    matrix_transpose(MATRIX_K, MATRIX_N, colB, B);
+    Tab *tmp = B;
+    B = colB;
+    free(tmp, q);
+  }
+
+  if constexpr (B_layout == layout::ext_intel_packed) {
+    Tab *vnniB = malloc_shared<Tab>(MATRIX_K * MATRIX_N, q);
+    matrix_vnni(MATRIX_K, MATRIX_N, B, vnniB, vnni_factor<Tab, B_layout>());
+    Tab *tmp = B;
+    B = vnniB;
+    free(tmp, q);
+  }
+
+  matrix_multiply<Tc, Tab, MATRIX_M, MATRIX_N, MATRIX_K, TM, TN, TK, A_layout,
+                  B_layout>(C, A, B, q);
+  assert(matrix_compare(MATRIX_M, MATRIX_N, C, D));
+  std::cout << "passed" << std::endl;
 
   free(A, q);
   free(B, q);
-  free(vnniB, q);
   free(C, q);
   free(D, q);
+}
 
-  return !res;
+template <layout A_layout, layout B_layout> void test_all() {
+  std::cout << "bf16: ";
+  test<bfloat16, float, /*MATRIX_M*/ 1024 + 20, /*MATRIX_N*/ 1024 + 20,
+       /*MATRIX_K*/ 1024 + 24, /*TM*/ 8, /*TN*/ 16, /*TK*/ 16, A_layout,
+       B_layout>();
+  std::cout << "half: ";
+  test<half, float, 1024 + 20, 1024 + 20, 1024 + 24, 8, 16, 16, A_layout,
+       B_layout>();
+  std::cout << "int8: ";
+  test<int8_t, int32_t, 1024, 1024 + 20, 1024 + 24, 8, 16, 32, A_layout,
+       B_layout>();
```

Comment:
I updated MATRIX_M, MATRIX_N and MATRIX_K to satisfy the requirement: Surface pitch (encoded_value + 1) must be a multiple of QW (8 bytes).